File: benchmarks/83-direct/direct.m
(* Experiment 83 -- DIRECT (DIviding RECTangles) global optimization
   (NMinimize, Method -> {"DIRECT", ...}).

   MEASURES src/numerical_calculus/nm_direct.c -- a deterministic Lipschitzian
   search that normalizes the box to the unit hypercube and repeatedly subdivides
   the "potentially optimal" cells (Jones, Perttunen & Stuckman 1993; locally-
   biased DIRECT-L of Gablonsky & Kelley 2001) -- against scipy.optimize.direct.
   Same problems, same labels, same order, scipy's default parameters on both
   sides (locally_biased True, eps 1e-4, maxiter 1000, maxfun 1000 n). DIRECT is
   deterministic: there is no seed. The race is raw-vs-raw -- scipy's direct does
   no local polish, so "PostProcess" -> False disables Mathilda's for an apples-
   to-apples comparison of the same algorithm. The check is the objective at the
   reported point rounded to 10^3: two raw DIRECT runs agree at the basin, not to
   machine precision, so the rounding is coarser than the polished-engine
   benchmarks. Objectives over several variables are built with Sum/Table over
   fresh global symbols and passed through Evaluate[] past NMinimize's argument
   handling. *)

Get["../harness.m"];

require[{"NMinimize"}];

(* n-D Ackley over fresh global symbols, precomputed out of the timed region. *)
ackley[n_] := With[{v = Table[Symbol["av" <> ToString[i]], {i, 1, n}]},
                   -20 Exp[-0.2 Sqrt[Sum[v[[i]]^2, {i, 1, n}]/n]]
                   - Exp[Sum[Cos[2 Pi v[[i]]], {i, 1, n}]/n] + 20 + E];
ackleyV[n_] := Table[Symbol["av" <> ToString[i]], {i, 1, n}];
ackleyB[n_] := And @@ Table[-5 <= Symbol["av" <> ToString[i]] <= 5, {i, 1, n}];

ac5 = ackley[5]; ac5v = ackleyV[5]; ac5b = ackleyB[5];

(* 2-D benchmarks (bare x, y so the objective hits the compiled machine-precision
   path -- the fair analogue of scipy's numpy objective). *)
bench["01 Himmelblau 2D", NMinimize[{(x^2+y-11)^2+(x+y^2-7)^2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"DIRECT","PostProcess"->False}];];
check["01 Himmelblau 2D", Round[10^3 First[NMinimize[{(x^2+y-11)^2+(x+y^2-7)^2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"DIRECT","PostProcess"->False}]]]];

bench["02 Booth 2D", NMinimize[{(x+2y-7)^2+(2x+y-5)^2, -10<=x<=10 && -10<=y<=10}, {x,y}, Method->{"DIRECT","PostProcess"->False}];];
check["02 Booth 2D", Round[10^3 First[NMinimize[{(x+2y-7)^2+(2x+y-5)^2, -10<=x<=10 && -10<=y<=10}, {x,y}, Method->{"DIRECT","PostProcess"->False}]]]];

bench["03 Beale 2D", NMinimize[{(1.5-x+x y)^2+(2.25-x+x y^2)^2+(2.625-x+x y^3)^2, -4.5<=x<=4.5 && -4.5<=y<=4.5}, {x,y}, Method->{"DIRECT","PostProcess"->False}];];
check["03 Beale 2D", Round[10^3 First[NMinimize[{(1.5-x+x y)^2+(2.25-x+x y^2)^2+(2.625-x+x y^3)^2, -4.5<=x<=4.5 && -4.5<=y<=4.5}, {x,y}, Method->{"DIRECT","PostProcess"->False}]]]];

bench["04 Rosenbrock 2D", NMinimize[{100(y-x^2)^2+(1-x)^2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"DIRECT","PostProcess"->False}];];
check["04 Rosenbrock 2D", Round[10^3 First[NMinimize[{100(y-x^2)^2+(1-x)^2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"DIRECT","PostProcess"->False}]]]];

bench["05 Six-hump camel 2D", NMinimize[{(4-2.1 x^2+x^4/3)x^2+x y+(-4+4 y^2)y^2, -3<=x<=3 && -2<=y<=2}, {x,y}, Method->{"DIRECT","PostProcess"->False}];];
check["05 Six-hump camel 2D", Round[10^3 First[NMinimize[{(4-2.1 x^2+x^4/3)x^2+x y+(-4+4 y^2)y^2, -3<=x<=3 && -2<=y<=2}, {x,y}, Method->{"DIRECT","PostProcess"->False}]]]];

bench["06 Ackley 2D", NMinimize[{-20 Exp[-0.2 Sqrt[(x^2+y^2)/2]] - Exp[(Cos[2 Pi x]+Cos[2 Pi y])/2] + 20 + E, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"DIRECT","PostProcess"->False}];];
check["06 Ackley 2D", Round[10^3 First[NMinimize[{-20 Exp[-0.2 Sqrt[(x^2+y^2)/2]] - Exp[(Cos[2 Pi x]+Cos[2 Pi y])/2] + 20 + E, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"DIRECT","PostProcess"->False}]]]];

(* n-D benchmark (fresh global symbols, injected with Evaluate[]). *)
bench["07 Ackley 5D", NMinimize[{Evaluate[ac5], Evaluate[ac5b]}, Evaluate[ac5v], Method->{"DIRECT","PostProcess"->False}];];
check["07 Ackley 5D", Round[10^3 First[NMinimize[{Evaluate[ac5], Evaluate[ac5b]}, Evaluate[ac5v], Method->{"DIRECT","PostProcess"->False}]]]];

(* Bare x, y (the 2-D form written out) for the compiled machine-precision path. *)
bench["08 Styblinski-Tang 2D", NMinimize[{(x^4-16 x^2+5 x + y^4-16 y^2+5 y)/2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"DIRECT","PostProcess"->False}];];
check["08 Styblinski-Tang 2D", Round[10^3 First[NMinimize[{(x^4-16 x^2+5 x + y^4-16 y^2+5 y)/2, -5<=x<=5 && -5<=y<=5}, {x,y}, Method->{"DIRECT","PostProcess"->False}]]]];

bench["09 Rastrigin 2D", NMinimize[{20 + x^2-10 Cos[2 Pi x] + y^2-10 Cos[2 Pi y], -5.12<=x<=5.12 && -5.12<=y<=5.12}, {x,y}, Method->{"DIRECT","PostProcess"->False}];];
check["09 Rastrigin 2D", Round[10^3 First[NMinimize[{20 + x^2-10 Cos[2 Pi x] + y^2-10 Cos[2 Pi y], -5.12<=x<=5.12 && -5.12<=y<=5.12}, {x,y}, Method->{"DIRECT","PostProcess"->False}]]]];

File: benchmarks/83-direct/direct.py
#!/usr/bin/env python3
"""Experiment 83 -- DIRECT (DIviding RECTangles) global optimization (scipy col).

Same 9 problems as ``direct.m``, same labels, same order, solved with
scipy.optimize.direct using scipy's DEFAULT parameters (locally_biased=True,
eps 1e-4, maxiter 1000, maxfun 1000*n). DIRECT is deterministic -- there is no
seed. The race is DIRECT-vs-direct on the same algorithm, raw on both sides
(scipy's direct does no local polish, so the .m column disables its polish with
"PostProcess" -> False for an apples-to-apples comparison). These are standard
bounded multimodal benchmarks -- Himmelblau, Booth, Beale, Rosenbrock, six-hump
camel, Ackley (2-D and 5-D), Styblinski-Tang, Rastrigin -- whose global basin
both systems reach. The check is the objective at the reported point rounded to
10^3 (coarser than the polished-engine benchmarks: two raw DIRECT runs agree at
the basin, not to machine precision), matching the .m ``Round[10^3 First[...]]``.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
from scipy.optimize import direct

from harness import bench, check, require

require(["scipy.optimize:direct"])


def _solve(fun, bounds):
    return direct(fun, bounds)


def _himmelblau(v):
    x, y = v
    return (x * x + y - 11.0) ** 2 + (x + y * y - 7.0) ** 2

def _booth(v):
    x, y = v
    return (x + 2 * y - 7.0) ** 2 + (2 * x + y - 5.0) ** 2

def _beale(v):
    x, y = v
    return ((1.5 - x + x * y) ** 2 + (2.25 - x + x * y ** 2) ** 2
            + (2.625 - x + x * y ** 3) ** 2)

def _rosenbrock(v):
    x, y = v
    return 100.0 * (y - x * x) ** 2 + (1.0 - x) ** 2

def _camel(v):
    x, y = v
    return (4 - 2.1 * x * x + x ** 4 / 3) * x * x + x * y + (-4 + 4 * y * y) * y * y

def _ackley(v):
    v = np.asarray(v)
    n = v.size
    return (-20.0 * np.exp(-0.2 * np.sqrt(np.sum(v * v) / n))
            - np.exp(np.sum(np.cos(2 * np.pi * v)) / n) + 20.0 + np.e)

def _styblinski(v):
    v = np.asarray(v)
    return float(np.sum(v ** 4 - 16 * v * v + 5 * v) / 2.0)

def _rastrigin(v):
    v = np.asarray(v)
    return 10.0 * v.size + np.sum(v * v - 10.0 * np.cos(2 * np.pi * v))


_c01 = lambda: _solve(_himmelblau,  [(-5, 5), (-5, 5)])
_c02 = lambda: _solve(_booth,       [(-10, 10), (-10, 10)])
_c03 = lambda: _solve(_beale,       [(-4.5, 4.5), (-4.5, 4.5)])
_c04 = lambda: _solve(_rosenbrock,  [(-5, 5), (-5, 5)])
_c05 = lambda: _solve(_camel,       [(-3, 3), (-2, 2)])
_c06 = lambda: _solve(_ackley,      [(-5, 5), (-5, 5)])
_c07 = lambda: _solve(_ackley,      [(-5, 5)] * 5)
_c08 = lambda: _solve(_styblinski,  [(-5, 5), (-5, 5)])
_c09 = lambda: _solve(_rastrigin,   [(-5.12, 5.12), (-5.12, 5.12)])


def _chk(res):
    return int(np.floor(1e3 * res.fun + 0.5))


bench("01 Himmelblau 2D", _c01)
check("01 Himmelblau 2D", _chk(_c01()))

bench("02 Booth 2D", _c02)
check("02 Booth 2D", _chk(_c02()))

bench("03 Beale 2D", _c03)
check("03 Beale 2D", _chk(_c03()))

bench("04 Rosenbrock 2D", _c04)
check("04 Rosenbrock 2D", _chk(_c04()))

bench("05 Six-hump camel 2D", _c05)
check("05 Six-hump camel 2D", _chk(_c05()))

bench("06 Ackley 2D", _c06)
check("06 Ackley 2D", _chk(_c06()))

bench("07 Ackley 5D", _c07)
check("07 Ackley 5D", _chk(_c07()))

bench("08 Styblinski-Tang 2D", _c08)
check("08 Styblinski-Tang 2D", _chk(_c08()))

bench("09 Rastrigin 2D", _c09)
check("09 Rastrigin 2D", _chk(_c09()))
